contributors/design-proposals/auth/svcacct-token-volume-source.md

```diff
@@ -0,0 +1,138 @@
+must be attested to by the orchestration system in a way that allows a third
+party to trust that an arbitrary container on the cluster is who it says it is.
+In addition, infrastructure running on top of Kubernetes needs a simple
+mechanism to communicate with the Kubernetes APIs and to provide more complex
+tooling. Finally, a significant set of security challenges are associated with
+storing service account tokens as secrets in Kubernetes and limiting the methods
+whereby malicious parties can get access to these tokens will reduce the risk of
+platform compromise.
+
+As a platform, Kubernetes should evolve to allow identity management systems to
+provide more powerful workload identity without breaking existing use cases, and
+provide a simple out of the box workload identity that is sufficient to cover
+the requirements of bootstrapping low-level infrastructure running on
+Kubernetes. We expect that other systems to cover the more advanced scenarios,
+and see this effort as necessary glue to allow more powerful systems to succeed.
+
+With this feature, we hope to provide a backwards compatible replacement for
+service account tokens that strengthens the security and improves the
+scalability of the platform.
+
+## Proposal
+
+Kubernetes should:
+
+1.  Allow the automatic service account token injection to be disabled via
+    configuration.
+1.  Allow automatic service account token creation to occur only when a user
+    explicitly requests.
+1.  Implement a ServiceAccountToken volume projection that that maintains a
+    service account token requested by the node from the TokenRequest API.
+
+### Token Volume Projection
+
+A new volume projection will be implemented with an API that closely matches the
+TokenRequest API.
+
+```go
+type ServiceAccountTokenProjection struct {
+  // Audiences are the intendend audiences of the token.
+  Audiences []string
+  // ExpirationSeconds is the requested duration of validity of the service
+  // account token
+  ExpirationSeconds *int64
+  // Path is the relative path of the file to project the token into.
+  Path string
+}
+```
+
+A volume plugin implemented in the kubelet will project a service account token
+sourced from the TokenReview API into volumes created from
+ProjectedVolumeSources. As the token approaches expiration, the kubelet volume
+plugin will proactively rotate the service account token.
+
+To replace the current service account token secrets, we also need to inject the
+clusters CA certificate bundle. Initially we will deploy to data in a configmap
+per-namespace and reference it using a ConfigMapProjection.
+
+This fixes one scalability issue with the current service account token
+deployment model where secret GETs are a large portion of overall apiserver
+traffic. However it doesn't solve the per-namespace replicated CA data which
+needs to be stored in ETCD.
+
+To fix that issue we can implement a PublicConfigMapProjection that exactly
```

Comment:
Another option: extend config map source to perform an authorization check on the service accounts behalf to fetch a cross namespace config map.  That removes the need for a special config map source and makes users who have this use case happier.

Authorization is two checks - node authorizer for the reference, and a corresponding acl check as the service account or the node (arguably I should be able to grant my service account permission to fetch something and let the node fetch it)

Reply:
The node authorizer can do that check on the behalf of the service account:

1. If referenced
2. And if service account has access if outside of namespace

Reply:
Authorization referring to secrets is tracked in https://github.com/kubernetes/kubernetes/issues/4957

Reply:
That issue is about to be closed due to inactivity.

Reply:
Jordan proved to me my idea wouldn't work because the node authorizer doesn't know the who.  I think his argument was to not put this in the API but make it be a kubelet detail.  I think it's pretty reasonable for the short term not to introduce a new API here but instead make it part of kube config.

Reply:
Although we still need a way to tie the ca info (the implicit service account data) to a pod that doesn't request the secret.  The API on ServiceAccountToken is token path specific, but without the CA the token is pretty useless, and in-cluster config is broken.

Reply:
Would the admission controller automatically inject a public config map projection?